keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+  - [Hold indexes as []int typed in successPolicy](#hold-indexes-as-int-typed-in-successpolicy)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [x] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP extends the Job API to allow setting conditions under which an Indexed Job can be declared as succeeded.
+
+## Motivation
+
+There are cases where a batch workload requires an indexed job that want to care 
+only about drivers (leaders) indexes in determining the success or failure of a Job,
+for example MPI and PyTorch etc. This is currently not possible because the indexed job 
+is marked as Completed only if all indexes are succeeded.
+
+Additionally, there are cases where a batch workload can be declared as succeeded if a number of
+pods succeed, regardless of the index. Sometimes the number needs
+because an indexed job author might not know how many indexes the job will have in advance
+when the above batch workload has an autoscale semantic.
+
+Some third-party frameworks have implemented success policy.
+
+- [Kubeflow Training Operator](https://www.kubeflow.org/docs/components/training)
+- [Flux Operator](https://flux-framework.org/flux-operator/)
+
+### Goals
+
+- allow to mark a job as a `Succeeded` even if some of the indexes fail.
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` mode.
```

Comment:
Do we foresee extending this API for non-indexed jobs? Was it already discussed? If so, what are the reasons for dropping non-indexed jobs, scoping, or something more?

I'm ok to scope the PR to indexed, but it seems the stories 2-4 are all applicable to non-indexed jobs, and the API seems to mostly fit for non-indexed jobs already. The only field that is indexed-specific is `OnSucceededPods.indexes`.

Reply:
Yes, we have a discussion about non-indexed jobs here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1225362635

Since we don't have effective use cases now, I put the non-indexed job support in this.

@mimowo WDYT?

Reply:
Sgtm, maybe just add a note somewhere, say in the notes section

Reply:
Done.